plugins/pay.c

```diff
@@ -512,57 +504,53 @@ static struct command_result *waitsendpay_error(struct command *cmd,
 		return forward_error(cmd, buf, error, pc);
 	}
 
+	if (time_after(time_now(), pc->stoptime)) {
+		return waitsendpay_expired(cmd, pc);
+	}
+
 	if (failcode & NODE) {
-		nodeidtok = json_delve(buf, error, ".data.erring_node");
-		if (!nodeidtok)
+		struct node_id id;
+		const char *idstr;
+
+		if (!json_scan(buf, error, "{data:{erring_node:%}}",
+			       JSON_SCAN(json_to_node_id, &id)))
 			plugin_err(cmd->plugin, "waitsendpay error no erring_node '%.*s'",
 				   error->end - error->start, buf + error->start);
-		node_err = true;
-	} else {
-		scidtok = json_delve(buf, error, ".data.erring_channel");
-		if (!scidtok)
-			plugin_err(cmd->plugin, "waitsendpay error no erring_channel '%.*s'",
-				   error->end - error->start, buf + error->start);
-		dirtok = json_delve(buf, error, ".data.erring_direction");
-		if (!dirtok)
-			plugin_err(cmd->plugin, "waitsendpay error no erring_direction '%.*s'",
-				   error->end - error->start, buf + error->start);
-	}
 
-	if (time_after(time_now(), pc->stoptime)) {
-		return waitsendpay_expired(cmd, pc);
-	}
+		/* FIXME: Keep as node_id, don't use strings. */
+		idstr = node_id_to_hexstr(tmpctx, &id);
 
-	if (node_err) {
 		/* If failure is in routehint part, try next one */
 		if (node_or_channel_in_routehint(pc->plugin, pc->current_routehint,
-						 buf + nodeidtok->start,
-						 nodeidtok->end - nodeidtok->start))
+						 idstr, strlen(idstr)))
 			return next_routehint(cmd, pc);
 
 		/* Otherwise, add erring channel to exclusion list. */
-		tal_arr_expand(&pc->excludes,
-			       tal_fmt(pc->excludes, "%.*s",
-			       nodeidtok->end - nodeidtok->start,
-			       buf + nodeidtok->start));
+		tal_arr_expand(&pc->excludes, tal_steal(pc->excludes, idstr));
 	} else {
+		struct short_channel_id scid;
+		u32 dir;
+		const char *scidstr = short_channel_id_to_str(tmpctx, &scid);
+
+		if (!json_scan(buf, error, "{data:{erring_channel:%,erring_direction:%}}",
+			       JSON_SCAN(json_to_short_channel_id, &scid),
+			       JSON_SCAN(json_to_number, &dir)))
+			plugin_err(cmd->plugin, "waitsendpay error no erring_channel/direction '%.*s'",
+				   error->end - error->start, buf + error->start);
```

Comment:
`json_tok_full_len` + `json_tok_full`.

Reply:
I simply moved the code, but I've fixed it.